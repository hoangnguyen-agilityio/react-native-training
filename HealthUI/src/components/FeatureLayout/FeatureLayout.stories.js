import { storiesOf } from '@storybook/react-native';
import React from 'react';
import { Text } from 'react-native';
import FeatureLayout from '.';
import SectionLayout from '../SectionLayout';
import WaterIcon from '../../assets/water.svg';
import HeartIcon from '../../assets/duotone.svg';

storiesOf('FeatureLayout', module)
  .addDecorator(getStory => <SectionLayout>{getStory()}</SectionLayout>)
  .add('Default', () => (
    <FeatureLayout
      title="Water"
      icon={<WaterIcon />}
      number="1.0"
      measure="liters">
      <Text>Test Feature Layout</Text>
    </FeatureLayout>
  ))
  .add('Dark layout', () => (
    <FeatureLayout
      title="Heart"
      icon={<HeartIcon />}
      number="105"
      measure="bmp"
      theme="dark">
      <Text style={{ color: '#FFFFFF' }}>Test Feature Layout</Text>
    </FeatureLayout>
  ));
