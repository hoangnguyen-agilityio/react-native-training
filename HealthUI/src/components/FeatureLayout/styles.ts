import { StyleSheet } from 'react-native';
import colors from '../../constants/colors';

const styles = StyleSheet.create({
  root: {
    borderWidth: 3,
    borderColor: '#EDEFF7',
    borderRadius: 20,
    overflow: 'hidden',
    justifyContent: 'space-between',
  },
  darkRoot: {
    backgroundColor: colors.purple,
    borderRadius: 20,
    justifyContent: 'space-between',
  },
  gradientWrapper: {
    flex: 1,
    justifyContent: 'space-between',
  },
  header: {
    padding: 12,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  footer: {
    padding: 12,
    flexDirection: 'row',
  },
  title: {
    fontFamily: 'SF-Pro-Display-Bold',
    fontSize: 16,
    color: '#53668E',
  },
  darkTitle: {
    fontFamily: 'SF-Pro-Display-Bold',
    fontSize: 16,
    color: colors.white,
  },
  number: {
    fontFamily: 'SF-Pro-Display-Bold',
    fontSize: 16,
    // color: colors.purple,
    lineHeight: 24,
  },
  darkNumber: {
    fontFamily: 'SF-Pro-Display-Medium',
    fontSize: 16,
    color: '#DDDAF1',
    lineHeight: 24,
  },
  measure: {
    fontFamily: 'SF-Pro-Display-Regular',
    fontSize: 12,
    // color: '#A274CD',
    marginLeft: 4,
    lineHeight: 22,
  },
  darkMeasure: {
    fontFamily: 'SF-Pro-Display-Regular',
    fontSize: 12,
    color: '#DDDAF1',
    marginLeft: 4,
    lineHeight: 22,
  },
});

export default styles;
