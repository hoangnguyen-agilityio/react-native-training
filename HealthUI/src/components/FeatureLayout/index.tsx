import React, { ReactNode } from 'react';
import { View, Text, TouchableOpacity } from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import GradientText from '../GradientText';
import styles from './styles';

interface Props {
  title: string;
  icon: ReactNode;
  children?: ReactNode;
  theme?: 'dark' | 'light';
  number?: string;
  measure?: string;
  customStyle?: object;
  onPress?: () => void;
  linearGradientColors?: string[];
}

const FeatureLayout = (props: Props) => {
  const {
    title,
    theme,
    icon,
    children = null,
    number,
    measure,
    customStyle = {},
    onPress = () => {},
    linearGradientColors = [],
  } = props;
  const rootStyle = theme === 'dark' ? styles.darkRoot : styles.root;
  const titleStyle = theme === 'dark' ? styles.darkTitle : styles.title;
  const numberStyle = theme === 'dark' ? styles.darkNumber : styles.number;
  const measureStyle = theme === 'dark' ? styles.darkMeasure : styles.measure;
  const footer =
    number && measure ? (
      theme === 'dark' ? (
        <View style={styles.footer}>
          <Text style={numberStyle}>{number}</Text>
          <Text style={measureStyle}> {measure}</Text>
        </View>
      ) : (
        <View style={styles.footer}>
          <GradientText>
            <Text style={numberStyle}>{number}</Text>
            <Text style={measureStyle}> {measure}</Text>
          </GradientText>
        </View>
      )
    ) : null;

  if (linearGradientColors.length) {
    return (
      <TouchableOpacity
        style={{ ...rootStyle, ...customStyle }}
        onPress={onPress}>
        <LinearGradient
          colors={linearGradientColors}
          style={styles.gradientWrapper}>
          <View style={styles.header}>
            <Text style={titleStyle}>{title}</Text>
            {icon}
          </View>
          {children}

          {footer}
        </LinearGradient>
      </TouchableOpacity>
    );
  }
  return (
    <TouchableOpacity
      style={{ ...rootStyle, ...customStyle }}
      onPress={onPress}>
      <View style={styles.header}>
        <Text style={titleStyle}>{title}</Text>
        {icon}
      </View>
      {children}

      {footer}
    </TouchableOpacity>
  );
};

export default FeatureLayout;
