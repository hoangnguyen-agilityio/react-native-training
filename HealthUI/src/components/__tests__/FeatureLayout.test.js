import React from 'react';
import { Text, View } from 'react-native';
import renderer from 'react-test-renderer';
import FeatureLayout from '../FeatureLayout';
import WaterIcon from '../../assets/water.svg';

test('renders correctly', () => {
  const tree = renderer
    .create(
      <FeatureLayout
        title="Water"
        icon={<WaterIcon />}
        number="1.0"
        measure="liters">
        <Text>Test Feature Layout</Text>
      </FeatureLayout>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('component elements', () => {
  const tree = renderer.create(
    <FeatureLayout
      title="Water"
      icon={<WaterIcon />}
      number="1.0"
      measure="liters">
      <Text>Test Feature Layout</Text>
    </FeatureLayout>,
  ).root;

  expect(tree.findAllByType(Text)).toHaveLength(4);
  expect(tree.findAllByType(WaterIcon)).toHaveLength(1);
});
