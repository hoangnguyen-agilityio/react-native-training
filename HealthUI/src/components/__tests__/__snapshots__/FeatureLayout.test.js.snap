// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "#EDEFF7",
      "borderRadius": 20,
      "borderWidth": 3,
      "justifyContent": "space-between",
      "opacity": 1,
      "overflow": "hidden",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 12,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#53668E",
          "fontFamily": "SF-Pro-Display-Bold",
          "fontSize": 16,
        }
      }
    >
      Water
    </Text>
    <SvgMock />
  </View>
  <Text>
    Test Feature Layout
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "padding": 12,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#5142AB",
          "fontFamily": "SF-Pro-Display-Bold",
          "fontSize": 16,
          "lineHeight": 24,
        }
      }
    >
      1.0
    </Text>
    <Text
      style={
        Object {
          "color": "#A274CD",
          "fontFamily": "SF-Pro-Display-Regular",
          "fontSize": 12,
          "lineHeight": 22,
          "marginLeft": 4,
        }
      }
    >
      liters
    </Text>
  </View>
</View>
`;
